### PR TITLE
Types cleanup

### PR DESCRIPTION
**Pull request by rbuckton**
Adds explicit interfaces for various syntax kinds, adding some additional restrictions to types whose definition is too broad.

This is needed to support #5595.

### CODE REVIEW

File: src/compiler/types.ts [hunk 1]
```diff
@@ -929,18 +1068,36 @@ namespace ts {
         _statementBrand: any;
     }
 
+    // @kind(SyntaxKind.EmptyStatement)
+    export interface EmptyStatement extends Statement { }
+
+    // @kind(SyntaxKind.DebuggerStatement)
+    export interface DebuggerStatement extends Statement { }
```
Comment (DanielRosenwasser) [inline src/compiler/types.ts]: Consider adding brands to these.

**Comment by DanielRosenwasser** [comment]
:+1:

File: src/compiler/types.ts (hunk 1 above)
Comment (rbuckton) [inline src/compiler/types.ts]: I'd rather not litter all of the interfaces with brands. They seem more valuable when disambiguating structurally similar nodes that may not be legally valid in certain parts of the tree (such as `LeftHandSideExpression` vs `Expression`).

If we ever end up with Enum type literals, I might change it to:
```ts
export interface EmptyStatement extends Statement {
    kind: SyntaxKind.EmptyStatement;
}
...
```

File: src/compiler/types.ts [hunk 2]
```diff
@@ -929,18 +1068,36 @@ namespace ts {
...
     }
 
+    // @kind(SyntaxKind.EmptyStatement)
+    export interface EmptyStatement extends Statement { }
+
+    // @kind(SyntaxKind.DebuggerStatement)
+    export interface DebuggerStatement extends Statement { }
+
+    // @kind(SyntaxKind.MissingDeclaration)
+    // @factoryhidden("name", true)
```
Comment (mhegazy) [inline src/compiler/types.ts]: what does facroryhidden and factoryParam mean?

**Comment by mhegazy** [comment]
:+1:

File: src/compiler/types.ts (hunk 2 above)
Comment (rbuckton) [inline src/compiler/types.ts]: Sorry, meant to remove those.

**Pull request merged**